main/src/ca/uwaterloo/flix/language/errors/SafetyError.scala

```diff
@@ -597,4 +597,44 @@ object SafetyError {
          |""".stripMargin
     )
   }
+
+  /**
+    * An error raised to indicate that a function marked with the `@Tailrec` annotation
+    * has at least non-tail-recursive function call.
+    *
+    * @param loc the location of the non-tail-recursive call.
+    */
+  case class NonTailRecursiveFunction(loc: SourceLocation) extends SafetyError {
+    override def summary: String = "Function annotated with @Tailrec must be tail recursive."
+
+    override def message(formatter: Formatter): String = {
+      import formatter._
+      s"""${line(kind, source.name)}
+         |>> Function annotated with @Tailrec must be tail recursive.
+         |
+         |${code(loc, "A recursive call in non-tail position occurs here.")}
+         |
+         |""".stripMargin
+    }
+
+    override def explain(formatter: Formatter): Option[String] = Some({
```

Comment:
Not sure if we should have this explanation, but I included it, so we can discuss it.

Reply:
In both the explanation and error message I think it is important to understand that tail recursion (despite the name) has nothing to do with tail calls!

In fact, the point is simply that if a function calls another function in tail position then the stack does not grow.